wp-includes/mysql/class-wp-mysql-token.php

```diff
@@ -24,6 +24,114 @@ public function get_name(): string {
+			 *
+			 * Despite looking similar, these rules are different from the C-style
+			 * string escaping, so we cannot use "strip(c)slashes()" in this case.
+			 *
+			 * @TODO: Handle NO_BACKSLASH_ESCAPES SQL mode.
+			 *
+			 * See: https://dev.mysql.com/doc/refman/8.4/en/string-literals.html
+			 */
+			$replacements = array(
+				/*
+				 * MySQL special character escape sequences.
+				 */
+				'\0'   => chr( 0 ),  // An ASCII NULL (X'00') character.
+				"\'"   => "'",       // A single quote (') character.
+				'\"'   => '"',       // A double quote (") character.
+				'\b'   => chr( 8 ),  // A backspace character.
+				'\n'   => "\n",      // A newline (linefeed) character.
+				'\r'   => "\r",      // A carriage return character.
+				'\t'   => "\t",      // A tab character.
+				'\Z'   => chr( 26 ), // An ASCII 26 (Control+Z) character.
+
+				/*
+				 * Normalize escaping of "%" and "_" characters.
+				 *
+				 * MySQL has unusual handling for "\%" and "\_" in all string literals.
+				 * While other sequences follow the C-style escaping ("\?" is "?", etc.),
+				 * "\%" resolves to "\%" and "\_" resolves to "\_" (unlike in C strings).
+				 *
+				 * This means that "\%" behaves like "\\%", and "\_" behaves like "\\_".
+				 * To preserve this behavior, we need to add a second backslash here.
+				 *
+				 * From https://dev.mysql.com/doc/refman/8.4/en/string-literals.html:
+				 *   > The \% and \_ sequences are used to search for literal instances
+				 *   > of % and _ in pattern-matching contexts where they would otherwise
+				 *   > be interpreted as wildcard characters. If you use \% or \_ outside
+				 *   > of pattern-matching contexts, they evaluate to the strings \% and
+				 *   > \_, not to % and _.
+				 */
+				'\%'   => '\\\\%',
+				'\_'   => '\\\\_',
+
+				/*
+				 * Preserve a double backslash as-is, so that the trailing backslash
+				 * is not consumed as the beginning of an escape sequence like "\n".
+				 *
+				 * Resolving "\\" to "\" will be handled in the next step, where all
+				 * other backslash-prefixed characters resolve to their literal values.
+				 */
+				'\\\\' => '\\\\',
+
+				/*
+				 * The bounding quotes can also be escaped by being doubled.
+				 */
+				$quote . $quote
+					=> $quote,
+			);
+
+			/*
+			 * Apply the replacements.
+			 *
+			 * It is important to use "strtr()" and not "str_replace()", because
+			 * "str_replace()" applies replacements one after another, modifying
+			 * intermediate changes rather than just the original string:
+			 *
+			 *   - str_replace( [ 'a', 'b' ], [ 'b', 'c' ], 'ab' ); // 'cc' (bad)
+			 *   - strtr( 'ab', [ 'a' => 'b', 'b' => 'c' ] );       // 'bc' (good)
+			 */
+			$value = strtr( $value, $replacements );
+
+			/*
+			 * A backslash with any other character represents the character itself.
+			 * That is, \x evaluates to x, \\ evaluates to \, and \🙂 evaluates to 🙂.
+			 */
+			$value = preg_replace( '/\\\\(.)/u', '$1', $value );
```

Comment:
Now I feel nostalgic. The first time I've read how a quadruple backslash evaluates to a single backslash in preg_* functions was about 20 years ago in a PHP4 book. I'm old. 👴 Can we either document this or express this in a different way? Perhaps `$escaped_backslash = preg_quote("\x5C");` and `preg_replace( '/'. $escaped_backslash .'(.)/u', '$1', $value );`? Maybe that's an overkill. Feel free to make any call here, I just want to make sure this gets brought up.

Reply:
Good point! I did that together with the other improvements in https://github.com/Automattic/sqlite-database-integration/pull/47/commits/931c82bdec4251c23338785adbf463934291eb7a.